meson.build

```diff
@@ -55,15 +56,20 @@ else
     endif
 endif
 
+if get_option('default_library') != 'static'
+    extra_args += '-DINI_SHARED_LIB'
+endif
```

Comment:
This is only necessary on Windows, and I *think* only with MSVC?, and in that case you need a bit more complicated of a check:
- if default_library is shared, pass that arg
- if default_library is both, error out with the following message: `error('default_library=both is not supported with MSVC')`
- if default_library is static, skip

Reply:
The problem here is that with default_library=both you build both static and shared, but there is no option to define *different* defines for each. And my understanding is MSVC is so... interestingly designed... that it downright breaks your library if you add a sharedlib declspec when building statically. This is not a problem with GCC or clang.

Reply:
> This is only necessary on Windows, and I think only with MSVC?

This is necessary on Windows, but it's nice to have on other platforms too, as it makes .so files cleaner and lighter.

Also, I'm not sure that erroring out is necessary, as you can actually build `both` libraries without issues even on Windows.

I check for `!= 'static'` instead of `== 'shared'` because while you can compile the static library when `-DINI_SHARED_LIB` is defined, you can't compile the shared one when the macro is absent.

Reply:
> This is necessary on Windows, but it's nice to have on other platforms too, as it makes .so files cleaner and lighter.

No, I mean the -D define is unnecessary on other platforms, because on `__GNUC__` (gcc/clang) you just unconditionally --without checking for `INI_SHARED_LIB` -- define an `__attribute__ ((visibility ("default")))`.

And thus, even without the define, your .so files are always cleaner and lighter. Because sane compiler.

> Edit2: so, as a `BothLibraries` object in Meson behaves like a `SharedLibrary` when used, defining `INI_SHARED_LIB` when the lib type is `both` or `shared` seems reasonable to me.

The issue is that it will also *install* a `StaticLibrary`, and then attempting to use that from other build systems which are configured to search for and prefer static libraries, may produce broken results.

Given that the default is shared, and defining `both` is something you'd need to opt into, the default behavior is fine anyway.

Reply:
> No, I mean the -D define is unnecessary on other platforms, because on `__GNUC__` (gcc/clang) you just unconditionally --without checking for `INI_SHARED_LIB` -- define an `__attribute__ ((visibility ("default")))`.

Oh, you're right, I copied the various `#if`s from another project of mine and didn't noticed that this was the case 😅️

> The issue is that it will also _install_ a `StaticLibrary`, and then attempting to use that from other build systems which are configured to search for and prefer static libraries, may produce broken results.

Yep, calling `error()` on Windows might be the only solution.

> (I am no expert on Windows or MSVC. But I've been told that it "does" produce broken results -- you end up building both a dll and a static library, and the dll works, and the static library doesn't.)

Awesome.